{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.8.12","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# Visualizing Sequence Similarity Networks (SSNs) using an internactive network\n\nNetworks have nodes and edges. A <b>node</b> is a connection point and <b>edges</b> are the connections between them. In this exercise each protein is a node and each edge indicates an evalue (expectation value) smaller than 10e-40.\n\n## Examining the BLAST output\n\n<font color=blue><b>STEP 1:</b></font> Let's first look at the output of our BLASTP search by double clicking the file in the file browser. ","metadata":{}},{"cell_type":"markdown","source":"***\nYou should see three columns, with each column separated by a tab, they aren't labeled, but we dicated them using the outfmt command in our blast search. \n\n## Creating a Dataframe from the BLAST output\n\nAn <b>API (Application Programming Interface</b> is a piece of software that allow applications to talk to each other. A <b>dataframe</b> is a popular API that resembles a spreadsheet. The BLAST output is a tab separated file, and we will use <b>pandas</b> - a \"powerful Python data analysis toolkit\" to read our file and convert it to a dataframe (i.e. spreadsheet).\n\n<font color=blue><b>STEP 2:</b></font> Edit the code to replace the <b>\\<<<your file here\\>>></b> with the BLAST output file path (remember it is in the files directory). Then run the code below to convert the BLASTe40_out file into a dataframe. Since our BLAST output did not contain any headers (column labels), we can add them in.\n","metadata":{}},{"cell_type":"code","source":"import pandas as pd # imports the pandas functions\n\nheaderList = ['source','target','evalue']\n\nblast_data_con = pd.read_csv('<<<your file here>>>', sep='\\t', header=None)  # reads the BLAST output and looks for 'tab' to separate the values\n\nblast_data_con.columns = ['source', 'target', 'evalue']  # assigns names to the columns\n\nblast_data_con    # show what is in the dataframe.","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note that the complete dataframe is not shown, but it shows the first few rows and also the last few rows, which we call <b>edges in a network.</b>\n***\n## Creating dataframes of edges and nodes\n\n## Removing duplicates and self-references from edges\n\nThe code below removes duplicates (e.g. if <font color=\"blue\">a</font> finds <font color=\"blue\">b</font> and <font color=\"blue\">b</font> finds <font color=\"blue\">a</font>, we only need to keep one of them) and self-references (e.g. remove all instances of <font color=\"blue\">a</font> finds <font color=\"blue\">a</font>).\n\nThe code is a bit complicated and uses another function called numpy. Briefly, the code uses pandas and the numpy.sort function to create another dataframe with only the duplicates. We then \"subtract\" the dataframe containing duplicates from the original dataframe.\n\nWe call the new dataframe \"edges\".\n\n<font color=blue><b>STEP 4:</b></font> Run the code below to create a dataframe of unique edges. We have named this dataframe \"edges\".","metadata":{}},{"cell_type":"code","source":"import numpy as np\n\ndf = blast_data_con       # we assign the variable df to our blast_data_con so we can retain the original dataframe.\n\n#remove duplicates\nm=pd.DataFrame(np.sort(df[['source','target']])).duplicated()\ndf = df[~m]\n\n#removes self-reference\ndf = df[df.source != df.target]\n\nedges = df    # this is a unique set of edges\nedges         # show us the edges dataframe","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Creating a unique list of nodes\n\nWe will use the numpy.unique function to read through the sources and targets in the dataframe and make a list (called uniq_list) of nodes.\n\nThen we will use pandas to convert this list into a simple dataframe of nodes.\n\n<font color=blue><b>STEP 5:</b></font> Run the code below to create dataframe of unique nodes.","metadata":{}},{"cell_type":"code","source":"uniq_list = np.unique(df[['source', 'target']].values)   # find the unique values and put them in a list\n\nnodes = pd.DataFrame(uniq_list, columns = ['id'])  # make a node dataframe with the column header id\n\nnodes     # show us the nodes dataframe","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"***\n\n## Visualizing nodes and edges\n\nYou now have a set of nodes and edges that you can visualize. We will import ipycytoscape, A Cytoscape widget for displaying interactive networks. You can find more about Cytoscape here: https://cytoscape.org/ and ipycytoscape here: https://github.com/cytoscape/ipycytoscape.\n \nThis code is borrowed and edited from https://github.com/joseberlines, who has done some neat work with ipycytoscape.\n\n<font color=blue><b>STEP 6:</b></font> Run the code box below to visualize the graph (this might take a few minutes and might take a few seconds to appear even after the asterisk disappears).\n\n","metadata":{}},{"cell_type":"code","source":"# There isn't a real need to edit any of this since it is just making the network graph.\n\nimport json                     # json stands for JavaScript Object Notation and ipycytoscape reads the data in this format\nimport ipycytoscape             # the widget to visualize interactive networks.\nfrom ipywidgets import Output\n\ndef transform_into_ipycytoscape(nodes_df,edges_df):\n    \n    nodes_dict = nodes_df.to_dict('records')    # converts the nodes to a dictionary\n    edges_dict = edges_df.to_dict('records')    # convertst the edges to a dictionary \n\n    # building nodes\n\n    data_keys = ['id']  #this is a list of keys in stations (nodes)\n    position_keys = ['position_x','position_y']\n    rest_keys = ['score','idInt','name','score','group','removed','selected','selectable','locked','grabbed'\n                 'grabbable']\n    \n    \n    nodes_graph_list=[] #an empty list for making the json-like? file\n    for node in nodes_dict: #iterating over each node\n        dict_node = {}\n        data_sub_dict = {'data':{el:node[el] for el in data_keys}}\n        rest_sub_dict = {el:node[el] for el in node.keys() if el in rest_keys}\n        posi_sub_dict = {}\n        if 'position_x' in node.keys() and 'position_y' in node.keys():\n            posi_sub_dict = {'position':{el:node[el] for el in node.keys() if el in position_keys}}\n        \n        dict_node = {**data_sub_dict,**rest_sub_dict,**posi_sub_dict}\n        nodes_graph_list.append(dict_node)\n    \n    # building edges\n    \n    data_keys  = ['source','target','evalue'] #this is a list of keys in edges\n    data_keys2 = ['label','classes'] \n    rest_keys  = ['score','weight','group','networkId','networkGroupId','intn','rIntnId','group','removed','selected','selectable','locked','grabbed','grabbable','classes']\n    position_keys = ['position_x','position_y']\n    \n    edges_graph_list = []\n    for edge in edges_dict:\n        dict_edge = {}\n        data_sub_dict = {el:edge[el] for el in data_keys}\n        data_sub_dict2 = {el:edge[el] for el in edge.keys() if el in data_keys2}\n        rest_sub_dict = {el:edge[el] for el in edge.keys() if el in rest_keys}\n        \n        dict_edge = {'data':{**data_sub_dict,**data_sub_dict2},**rest_sub_dict}\n        edges_graph_list.append(dict_edge)\n    \n    #print(edges_graph_list)\n    \n    total_graph_dict = {'nodes': nodes_graph_list, 'edges':edges_graph_list}\n    \n    #print(total_graph_dict)\n    \n    # building the style\n    all_node_style = ['background-color','background-opacity',\n                     'font-family','font-size','label','width',\n                     'shape','height','width','text-valign','text-halign']\n    all_edge_style = ['background-color','background-opacity',\n                     'font-family','font-size','label','width','line-color', \n                     ]\n    \n    total_style_dict = {}\n    style_elements=[]\n    for node in nodes_dict:\n        node_dict = {'selector': f'node[id = \\\"{node[\"id\"]}\\\"]'}\n        style_dict ={\"style\": { el:node[el] for el in node.keys() if el in all_node_style}}\n        node_dict.update(style_dict)\n        style_elements.append(node_dict)\n    \n    for edge in edges_dict:\n        edge_dict = {'selector': f'edge[id = \\\"{edge[\"source\"]}\\\"]'}\n        style_dict ={\"style\": { el:edge[el] for el in edge.keys() if el in all_edge_style}}\n        edge_dict.update(style_dict)\n        style_elements.append(edge_dict)\n    \n    # the graph\n    data_graph = json.dumps(total_graph_dict)\n    json_to_python = json.loads(data_graph)\n    result_cyto = ipycytoscape.CytoscapeWidget()\n    result_cyto.graph.add_graph_from_json(json_to_python)    \n    result_cyto.set_style(style_elements)\n    result_cyto.set_layout(name='grid')   #concentric, cola, or grid\n\n    out = Output()\n\n    \"\"\"\n    def log_clicks(node):\n        with out:\n            print(f'clicked: {(node)}')\n\n    #def log_mouseovers(node):\n        #with out:\n            #print(f'mouseover: {pformat(node)}')\n\n    result_cyto.on('node', 'click', log_clicks)\n    #result_cyto.on('node', 'mouseover', log_mouseovers)\n    \n    \"\"\"  \n    return result_cyto, out\n\nnetwork, out =transform_into_ipycytoscape(nodes,edges)\ndisplay(network)\ndisplay(out)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"<font color=blue><b>STEP 7:</b></font> Try zooming in and out of the network. You can also grab and move nodes.\n\nWhile we did make an interactive network of nodes and edges, the problem is that this network doesn't contain enough information to give us insight into the connections among histone proteins.\n\n<font color=blue><b>STEP 8:</b></font> The code below adds some information to the dataframe for our nodes. It includes color and labels and sets a default size for the nodes. Run the code below to update the dataframe and then run the code to visualize the network again. In case you would like to see the full range of named colors, check out this site: https://matplotlib.org/stable/gallery/color/named_colors.html.\n","metadata":{}},{"cell_type":"code","source":"# Let's add some new columns to our dataframe\nnodes['label'] = nodes['id']        # creates a label using the id\nnodes['background-color']='cyan'    # our default color is cyan, but could be anything.\nnodes['width']='24'                \nnodes['height']='24'\nnodes['font-size']='20'\nnodes['text-valign']='center'\nnodes['text-halign']='center'\nnodes['count'] = '1'\n\nnodes","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"***\nThe network is still very difficult to use. Knowledge of the number of connections might help us to find nodes that make many connections and nodes that make only a few connections easily.\n\n<font color=blue><b>STEP 9:</b></font> The code below creates a list using our node IDs. It goes through the list and counts the number of edges to targets. Then it changes the variable count in the dataframe for that item. Lastly, the height and width are determined using the size * 10. I just made that up, you could try any multiple or even an exponent (e.g. \\**2).\n","metadata":{}},{"cell_type":"code","source":"#Let's change the size of the nodes based on the number of connections. \ncol_one_list = nodes['id'].tolist()     # make a list from the dataframe\n\nfor item in col_one_list: \n    size = len(edges[edges['source']==item]) + len(edges[edges['target']==item])\n    nodes.loc[nodes['id'] == item, 'count']=size\n    size = size*10\n    nodes.loc[nodes['id'] == item, 'width']=size\n    nodes.loc[nodes['id'] == item, 'height']=size\n\nnodes['font-size']='50'   # let's also increase the size of the font here.\n    \nnodes","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"This is finally starting to give us some information. \n\n<font color=blue><b>STEP 10:</b></font> Let's try another last amendment to our dataframe by giving a color to our knowns. Note that these are the IDs in the \"known_histone.fasta\" file, and I just picked a different color for each. Run the code below and then rerun the network visualization.","metadata":{}},{"cell_type":"code","source":"\n#Here we can assign colors to nodes that connect to one of our knowns!\n\nnodes.loc[nodes['id'] == '1U8R_IDER','background-color']  = 'red'\nnodes.loc[nodes['id'] == '1C0W_DTXR','background-color']  = 'orange'\nnodes.loc[nodes['id'] == '6O5C_MTSR','background-color']  = 'yellow'\nnodes.loc[nodes['id'] == '3HRT_SCAR','background-color']  = 'green'\nnodes.loc[nodes['id'] == '5CVI_SLOR','background-color']  = 'blue'\nnodes.loc[nodes['id'] == '3R60_MNTR','background-color']  = 'magenta'\n\nnodes","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"\nLet's see if adding more color can help us to generate clusters.\n\n<font color=blue><b>STEP 12:</b></font> In this step, if a node is identified by a known, we will color it the same as the known. Note that the last section recolors the known nodes just in case they were identified by another known directly. Run the code below and then rerun the network visualization.","metadata":{}},{"cell_type":"code","source":"records = edges.to_records(index=False)\nresult = list(records)\n\nfor item in result:\n    #print(item)\n    if item[1] == '1U8R_IDER':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'red'\n    if item[1] == '5CVI_SLOR':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'blue'\n    if item[1] == '3HRT_SCAR':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'green'\n    if item[1] == '1C0W_DTXR':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'orange'\n    if item[1] == '3R60_MNTR':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'magenta'\n    if item[1] == '6O5C_MTSR':\n        nodes.loc[nodes['id'] == item[0],'background-color']  = 'yellow'\n\nnodes.loc[nodes['id'] == '1U8R_IDER','background-color']  = 'red'\nnodes.loc[nodes['id'] == '1C0W_DTXR','background-color']  = 'orange'\nnodes.loc[nodes['id'] == '6O5C_MTSR','background-color']  = 'yellow'\nnodes.loc[nodes['id'] == '3HRT_SCAR','background-color']  = 'green'\nnodes.loc[nodes['id'] == '5CVI_SLOR','background-color']  = 'blue'\nnodes.loc[nodes['id'] == '3R60_MNTR','background-color']  = 'magenta'\n        \nnodes","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"***\n\n<font color=blue><b>STEP 13:</b></font> Interact with the graph to create clusters. Briefly, cluster by first moving the larger known clusters to the outside of the grid graph. Then bring the similarly colored groups together. Finally, arrange the remaining cyan nodes nearer to their connections as dictated by the edges.\n\n<font color=blue><b>STEP 14:</b></font> Use your clustered graph to answer the following questions:\n\n    1. Which histone proteins are most closely related to each other? \n    2. Where do the A. vaga histones cluster?\n    3. What sequence (or sequences) connect the 5CVI_SLOR protein to the 1C0W_DTXR protein?","metadata":{}},{"cell_type":"markdown","source":"<font color=blue><b>STEP 15:</b></font> <b>Challenge Question:</b> Having identified the sequence links between 5CVI_SLOR protein and 1C0W_DTXR protein (Step 14 question 3), create a fasta file that contains these linking sequences and the knowns (in dtxr_pdbs.fasta). Then create a multiple alignment from this fasta. Using the mutliple alignment output, assign functions to the linking sequences - choices are: DtxR/IdeR-like, SloR/ScaR-like, MtsR-like or unknown function.   ","metadata":{}},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}